File: src/core/SimulationScheduler.js
import { CLOCK_POLICIES } from './GameState.js';

export const SIMULATION_CLOCKS = Object.freeze({
  RENDER: 'RENDER',
  GAMEPLAY: 'GAMEPLAY_REAL_TIME',
  PHYSICS: 'PHYSICS_FIXED',
  CITY: 'CITY_LOGICAL',
  UI: 'UI',
  PAUSED: 'PAUSED'
});

export const SIMULATION_STAGES = Object.freeze({
  INPUT: 'INPUT',
  FIXED_PHYSICS: 'FIXED_PHYSICS',
  GAMEPLAY: 'GAMEPLAY',
  CITY: 'CITY',
  PRESENTATION: 'PRESENTATION',
  CAMERA: 'CAMERA',
  RENDER: 'RENDER'
});

export const SIMULATION_STAGE_ORDER = Object.freeze([
  SIMULATION_STAGES.INPUT,
  SIMULATION_STAGES.FIXED_PHYSICS,
  SIMULATION_STAGES.GAMEPLAY,
  SIMULATION_STAGES.CITY,
  SIMULATION_STAGES.PRESENTATION,
  SIMULATION_STAGES.CAMERA,
  SIMULATION_STAGES.RENDER
]);

const STAGE_CLOCKS = Object.freeze({
  [SIMULATION_STAGES.INPUT]: SIMULATION_CLOCKS.UI,
  [SIMULATION_STAGES.FIXED_PHYSICS]: SIMULATION_CLOCKS.PHYSICS,
  [SIMULATION_STAGES.GAMEPLAY]: SIMULATION_CLOCKS.GAMEPLAY,
  [SIMULATION_STAGES.CITY]: SIMULATION_CLOCKS.CITY,
  [SIMULATION_STAGES.PRESENTATION]: SIMULATION_CLOCKS.UI,
  [SIMULATION_STAGES.CAMERA]: SIMULATION_CLOCKS.RENDER,
  [SIMULATION_STAGES.RENDER]: SIMULATION_CLOCKS.RENDER
});

const ACTIVE_GAMEPLAY_POLICIES = new Set([
  CLOCK_POLICIES.CITY,
  CLOCK_POLICIES.BUILDER,
  CLOCK_POLICIES.STREET
]);

const ACTIVE_CITY_POLICIES = new Set([
  CLOCK_POLICIES.CITY,
  CLOCK_POLICIES.BUILDER,
  CLOCK_POLICIES.STREET
]);

const MULTIPLIED_CITY_POLICIES = new Set([
  CLOCK_POLICIES.CITY,
  CLOCK_POLICIES.BUILDER
]);

const VALID_CLOCK_POLICIES = new Set(Object.values(CLOCK_POLICIES));
const VALID_STAGES = new Set(SIMULATION_STAGE_ORDER);
const EPSILON = 1e-9;

function assertPositiveFinite(value, label) {
  if (!Number.isFinite(value) || value <= 0) {
    throw new RangeError(`${label} must be a positive finite number`);
  }
}

function normalizeDelta(value) {
  return Number.isFinite(value) ? Math.max(0, value) : 0;
}

function createClock() {
  return { delta: 0, elapsed: 0, ticks: 0 };
}

function snapshotClock(clock) {
  return Object.freeze({
    delta: clock.delta,
    elapsed: clock.elapsed,
    ticks: clock.ticks
  });
}

/**
 * Renderer-independent owner of MetroPulse's frame policy and clock domains.
 *
 * The scheduler intentionally knows nothing about Three.js, cannon-es, the DOM,
 * or individual game systems. The composition root registers small tasks in a
 * documented stage; the scheduler supplies timing, ordering, and state gates.
 */
export class SimulationScheduler {
  constructor({
    fixedPhysicsStep = 1 / 120,
    cityStep = 1,
    maxFrameDelta = 0.1,
    maxPhysicsStepsPerFrame = 10,
    maxCityTicksPerFrame = 30,
    getCityTimeScale = () => 1,
    initialClockPolicy = CLOCK_POLICIES.STOPPED
  } = {}) {
    assertPositiveFinite(fixedPhysicsStep, 'fixedPhysicsStep');
    assertPositiveFinite(cityStep, 'cityStep');
    assertPositiveFinite(maxFrameDelta, 'maxFrameDelta');
    if (!Number.isInteger(maxPhysicsStepsPerFrame) || maxPhysicsStepsPerFrame < 1) {
      throw new RangeError('maxPhysicsStepsPerFrame must be a positive integer');
    }
    if (!Number.isInteger(maxCityTicksPerFrame) || maxCityTicksPerFrame < 1) {
      throw new RangeError('maxCityTicksPerFrame must be a positive integer');
    }
    if (typeof getCityTimeScale !== 'function') {
      throw new TypeError('getCityTimeScale must be a function');
    }

    this.fixedPhysicsStep = fixedPhysicsStep;
    this.cityStep = cityStep;
    this.maxFrameDelta = maxFrameDelta;
    this.maxPhysicsStepsPerFrame = maxPhysicsStepsPerFrame;
    this.maxCityTicksPerFrame = maxCityTicksPerFrame;
    this.getCityTimeScale = getCityTimeScale;
    this.clockPolicy = CLOCK_POLICIES.STOPPED;
    this.tasks = new Map(SIMULATION_STAGE_ORDER.map(stage => [stage, []]));
    this.taskIds = new Set();
    this.registrationSerial = 0;
    this.frame = 0;
    this.lastTimestampMs = null;
    this.physicsAccumulator = 0;
    this.cityAccumulator = 0;
    this.lastFrameStats = Object.freeze({ physicsSteps: 0, cityTicks: 0, cityTimeScale: 0 });
    this.clocks = Object.fromEntries(
      Object.values(SIMULATION_CLOCKS).map(clock => [clock, createClock()])
    );
    this.setClockPolicy(initialClockPolicy);
  }

  get paused() {
    return this.clockPolicy === CLOCK_POLICIES.PAUSED;
  }

  get gameplayActive() {
    return ACTIVE_GAMEPLAY_POLICIES.has(this.clockPolicy);
  }

  get cityActive() {
    return ACTIVE_CITY_POLICIES.has(this.clockPolicy);
  }

  setClockPolicy(clockPolicy) {
    if (!VALID_CLOCK_POLICIES.has(clockPolicy)) {
      throw new RangeError(`Unknown simulation clock policy: ${String(clockPolicy)}`);
    }
    this.clockPolicy = clockPolicy;
    return this.clockPolicy;
  }

  registerTask({ id, stage, update, order = 0, enabled = null }) {
    if (typeof id !== 'string' || id.trim() === '') {
      throw new TypeError('task id must be a non-empty string');
    }
    if (this.taskIds.has(id)) throw new Error(`Simulation task already registered: ${id}`);
    if (!VALID_STAGES.has(stage)) throw new RangeError(`Unknown simulation stage: ${String(stage)}`);
    if (typeof update !== 'function') throw new TypeError('task update must be a function');
    if (!Number.isFinite(order)) throw new TypeError('task order must be finite');
    if (enabled !== null && typeof enabled !== 'function') {
      throw new TypeError('task enabled predicate must be a function or null');
    }

    const task = {
      id,
      stage,
      update,
      order,
      enabled,
      serial: this.registrationSerial++
    };
    const stageTasks = this.tasks.get(stage);
    stageTasks.push(task);
    stageTasks.sort((left, right) => left.order - right.order || left.serial - right.serial);
    this.taskIds.add(id);

    let registered = true;
    return () => {
      if (!registered) return false;
      registered = false;
      this.taskIds.delete(id);
      const index = stageTasks.indexOf(task);
      if (index >= 0) stageTasks.splice(index, 1);
      return index >= 0;
    };
  }

  getCityTimeScaleForPolicy() {
    if (!this.cityActive) return 0;
    const requestedScale = Number(this.getCityTimeScale({
      clockPolicy: this.clockPolicy,
      frame: this.frame
    }));
    const safeScale = Number.isFinite(requestedScale) ? Math.max(0, requestedScale) : 0;
    return MULTIPLIED_CITY_POLICIES.has(this.clockPolicy)
      ? safeScale
      : (safeScale > 0 ? 1 : 0);
  }

  runFrame(timestampMs) {
    const timestamp = Number(timestampMs);
    if (!Number.isFinite(timestamp)) return this.#advanceFrame(0);
    if (this.lastTimestampMs === null) {
      this.lastTimestampMs = timestamp;
      return this.#advanceFrame(0);
    }
    const deltaSeconds = Math.max(0, (timestamp - this.lastTimestampMs) / 1000);
    this.lastTimestampMs = timestamp;
    return this.#advanceFrame(deltaSeconds);
  }

  resetFrameClock(timestampMs = null) {
    this.lastTimestampMs = Number.isFinite(timestampMs) ? Number(timestampMs) : null;
  }

  advanceFrame(realDeltaSeconds) {
    this.#advanceFrame(realDeltaSeconds);
    return this.snapshot();
  }

  #advanceFrame(realDeltaSeconds) {
    const renderDelta = normalizeDelta(realDeltaSeconds);
    const boundedDelta = Math.min(renderDelta, this.maxFrameDelta);
    this.frame += 1;

    this.#advanceClock(SIMULATION_CLOCKS.RENDER, renderDelta);
    this.#advanceClock(SIMULATION_CLOCKS.UI, boundedDelta);

    const frameContext = {
      frame: this.frame,
      clockPolicy: this.clockPolicy,
      renderDelta,
      boundedDelta,
      gameplayDelta: 0,
      cityTimeScale: 0,
      physicsAlpha: 0,
      clocks: this.clocks
    };

    this.#runStage(SIMULATION_STAGES.INPUT, boundedDelta, frameContext);

    // Input can synchronously request a state transition. Re-read the policy
    // here so pausing cannot leak one final physics/gameplay/city update.
    const gameplayDelta = this.gameplayActive ? boundedDelta : 0;
    const cityTimeScale = this.getCityTimeScaleForPolicy();
    frameContext.clockPolicy = this.clockPolicy;
    frameContext.gameplayDelta = gameplayDelta;
    frameContext.cityTimeScale = cityTimeScale;
    this.#advanceClock(SIMULATION_CLOCKS.GAMEPLAY, gameplayDelta);
    this.#advanceClock(SIMULATION_CLOCKS.PAUSED, this.paused ? renderDelta : 0);
    this.clocks[SIMULATION_CLOCKS.PHYSICS].delta = 0;
    this.clocks[SIMULATION_CLOCKS.CITY].delta = 0;

    let physicsSteps = 0;
    if (gameplayDelta > 0) {
      this.physicsAccumulator += gameplayDelta;
      while (
        this.physicsAccumulator + EPSILON >= this.fixedPhysicsStep
        && physicsSteps < this.maxPhysicsStepsPerFrame
      ) {
        this.#advanceClock(SIMULATION_CLOCKS.PHYSICS, this.fixedPhysicsStep);
        this.#runStage(SIMULATION_STAGES.FIXED_PHYSICS, this.fixedPhysicsStep, frameContext);
        this.physicsAccumulator -= this.fixedPhysicsStep;
        if (Math.abs(this.physicsAccumulator) < EPSILON) this.physicsAccumulator = 0;
        physicsSteps += 1;
      }
    }
    frameContext.physicsAlpha = Math.min(1, this.physicsAccumulator / this.fixedPhysicsStep);

    if (gameplayDelta > 0) {
      this.#runStage(SIMULATION_STAGES.GAMEPLAY, gameplayDelta, frameContext);
    }

    let cityTicks = 0;
    const cityDelta = gameplayDelta * cityTimeScale;
    if (cityDelta > 0) {
      this.cityAccumulator += cityDelta;
      while (
        this.cityAccumulator + EPSILON >= this.cityStep
        && cityTicks < this.maxCityTicksPerFrame
      ) {
        this.#advanceClock(SIMULATION_CLOCKS.CITY, this.cityStep);
        this.#runStage(SIMULATION_STAGES.CITY, this.cityStep, frameContext);
        this.cityAccumulator -= this.cityStep;
        if (Math.abs(this.cityAccumulator) < EPSILON) this.cityAccumulator = 0;
        cityTicks += 1;
      }
    }

    this.#runStage(SIMULATION_STAGES.PRESENTATION, boundedDelta, frameContext);
    this.#runStage(SIMULATION_STAGES.CAMERA, boundedDelta, frameContext);
    this.#runStage(SIMULATION_STAGES.RENDER, renderDelta, frameContext);

    this.lastFrameStats = Object.freeze({ physicsSteps, cityTicks, cityTimeScale });
    return this.lastFrameStats;
  }

  #advanceClock(clockName, delta) {
    const clock = this.clocks[clockName];
    clock.delta = delta;
    if (delta <= 0) return;
    clock.elapsed += delta;
    clock.ticks += 1;
  }

  #runStage(stage, delta, frameContext) {
    frameContext.stage = stage;
    frameContext.clock = STAGE_CLOCKS[stage];
    frameContext.delta = delta;
    for (const task of this.tasks.get(stage)) {
      if (task.enabled && !task.enabled(frameContext)) continue;
      task.update(delta, frameContext);
    }
  }

  snapshot() {
    return Object.freeze({
      frame: this.frame,
      clockPolicy: this.clockPolicy,
      paused: this.paused,
      gameplayActive: this.gameplayActive,
      cityActive: this.cityActive,
      fixedPhysicsStep: this.fixedPhysicsStep,
      cityStep: this.cityStep,
      physicsRemainder: this.physicsAccumulator,
      cityRemainder: this.cityAccumulator,
      lastFrame: this.lastFrameStats,
      clocks: Object.freeze(Object.fromEntries(
        Object.entries(this.clocks).map(([name, clock]) => [name, snapshotClock(clock)])
      ))
    });
  }
}

export default SimulationScheduler;

File: godot/MetroPulse.Domain/Core/SimulationScheduler.cs
using System.Collections.ObjectModel;
using System.Text.Json.Serialization;

namespace MetroPulse.Domain.Core;

public enum SimulationStage
{
    Input,
    FixedPhysics,
    Gameplay,
    City,
    Presentation,
    Camera,
    Render,
}

public static class SimulationScheduleCatalog
{
    public const string RenderClock = "RENDER";
    public const string GameplayClock = "GAMEPLAY_REAL_TIME";
    public const string PhysicsClock = "PHYSICS_FIXED";
    public const string CityClock = "CITY_LOGICAL";
    public const string UiClock = "UI";
    public const string PausedClock = "PAUSED";

    public static readonly IReadOnlyList<string> Clocks = Array.AsReadOnly([
        RenderClock,
        GameplayClock,
        PhysicsClock,
        CityClock,
        UiClock,
        PausedClock,
    ]);

    public static readonly IReadOnlyList<SimulationStage> StageOrder = Array.AsReadOnly([
        SimulationStage.Input,
        SimulationStage.FixedPhysics,
        SimulationStage.Gameplay,
        SimulationStage.City,
        SimulationStage.Presentation,
        SimulationStage.Camera,
        SimulationStage.Render,
    ]);

    public static string ToToken(this SimulationStage stage) => stage switch
    {
        SimulationStage.FixedPhysics => "FIXED_PHYSICS",
        _ => stage.ToString().ToUpperInvariant(),
    };

    public static string GetClock(SimulationStage stage) => stage switch
    {
        SimulationStage.Input or SimulationStage.Presentation => UiClock,
        SimulationStage.FixedPhysics => PhysicsClock,
        SimulationStage.Gameplay => GameplayClock,
        SimulationStage.City => CityClock,
        SimulationStage.Camera or SimulationStage.Render => RenderClock,
        _ => throw new ArgumentOutOfRangeException(nameof(stage), stage, null),
    };
}

public sealed record SimulationClockSnapshot
{
    [JsonPropertyName("delta")]
    public required double Delta { get; init; }

    [JsonPropertyName("elapsed")]
    public required double Elapsed { get; init; }

    [JsonPropertyName("ticks")]
    public required long Ticks { get; init; }
}

public sealed record SimulationFrameStats
{
    [JsonPropertyName("physicsSteps")]
    public required int PhysicsSteps { get; init; }

    [JsonPropertyName("cityTicks")]
    public required int CityTicks { get; init; }

    [JsonPropertyName("cityTimeScale")]
    public required double CityTimeScale { get; init; }
}

public sealed record SimulationSchedulerSnapshot
{
    [JsonPropertyName("frame")]
    public required long Frame { get; init; }

    [JsonPropertyName("clockPolicy")]
    public required string ClockPolicy { get; init; }

    [JsonPropertyName("paused")]
    public required bool Paused { get; init; }

    [JsonPropertyName("gameplayActive")]
    public required bool GameplayActive { get; init; }

    [JsonPropertyName("cityActive")]
    public required bool CityActive { get; init; }

    [JsonPropertyName("fixedPhysicsStep")]
    public required double FixedPhysicsStep { get; init; }

    [JsonPropertyName("cityStep")]
    public required double CityStep { get; init; }

    [JsonPropertyName("physicsRemainder")]
    public required double PhysicsRemainder { get; init; }

    [JsonPropertyName("cityRemainder")]
    public required double CityRemainder { get; init; }

    [JsonPropertyName("lastFrame")]
    public required SimulationFrameStats LastFrame { get; init; }

    [JsonPropertyName("clocks")]
    public required IReadOnlyDictionary<string, SimulationClockSnapshot> Clocks { get; init; }
}

public sealed class SimulationFrameContext
{
    private readonly Func<IReadOnlyDictionary<string, SimulationClockSnapshot>> clocksProvider;

    internal SimulationFrameContext(Func<IReadOnlyDictionary<string, SimulationClockSnapshot>> clocksProvider) =>
        this.clocksProvider = clocksProvider;

    public long Frame { get; internal set; }

    public ClockPolicy ClockPolicy { get; internal set; }

    public double RenderDelta { get; internal set; }

    public double BoundedDelta { get; internal set; }

    public double GameplayDelta { get; internal set; }

    public double CityTimeScale { get; internal set; }

    public double PhysicsAlpha { get; internal set; }

    public SimulationStage Stage { get; internal set; }

    public string Clock { get; internal set; } = SimulationScheduleCatalog.UiClock;

    public double Delta { get; internal set; }

    public IReadOnlyDictionary<string, SimulationClockSnapshot> Clocks => clocksProvider();
}

public sealed class SimulationScheduler
{
    private const double Epsilon = 1e-9;
    private static readonly HashSet<ClockPolicy> ActiveGameplayPolicies =
        [ClockPolicy.City, ClockPolicy.Builder, ClockPolicy.Street];
    private static readonly HashSet<ClockPolicy> ActiveCityPolicies =
        [ClockPolicy.City, ClockPolicy.Builder, ClockPolicy.Street];
    private static readonly HashSet<ClockPolicy> MultipliedCityPolicies =
        [ClockPolicy.City, ClockPolicy.Builder];

    private readonly Dictionary<SimulationStage, List<ScheduledTask>> tasks =
        SimulationScheduleCatalog.StageOrder.ToDictionary(stage => stage, _ => new List<ScheduledTask>());
    private readonly HashSet<string> taskIds = new(StringComparer.Ordinal);
    private readonly Dictionary<string, MutableClock> clocks =
        SimulationScheduleCatalog.Clocks.ToDictionary(name => name, _ => new MutableClock(), StringComparer.Ordinal);
    private readonly Func<ClockScaleContext, double> getCityTimeScale;
    private long registrationSerial;
    private double? lastTimestampMs;
    private double physicsAccumulator;
    private double cityAccumulator;

    public SimulationScheduler(
        double fixedPhysicsStep = 1d / 120d,
        double cityStep = 1,
        double maxFrameDelta = 0.1,
        int maxPhysicsStepsPerFrame = 10,
        int maxCityTicksPerFrame = 30,
        Func<ClockScaleContext, double>? getCityTimeScale = null,
        ClockPolicy initialClockPolicy = ClockPolicy.Stopped)
    {
        FixedPhysicsStep = RequirePositive(fixedPhysicsStep, nameof(fixedPhysicsStep));
        CityStep = RequirePositive(cityStep, nameof(cityStep));
        MaxFrameDelta = RequirePositive(maxFrameDelta, nameof(maxFrameDelta));
        if (maxPhysicsStepsPerFrame < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxPhysicsStepsPerFrame));
        }
        if (maxCityTicksPerFrame < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxCityTicksPerFrame));
        }
        MaxPhysicsStepsPerFrame = maxPhysicsStepsPerFrame;
        MaxCityTicksPerFrame = maxCityTicksPerFrame;
        this.getCityTimeScale = getCityTimeScale ?? (_ => 1);
        SetClockPolicy(initialClockPolicy);
    }

    public double FixedPhysicsStep { get; }

    public double CityStep { get; }

    public double MaxFrameDelta { get; }

    public int MaxPhysicsStepsPerFrame { get; }

    public int MaxCityTicksPerFrame { get; }

    public long Frame { get; private set; }

    public ClockPolicy ClockPolicy { get; private set; }

    public bool Paused => ClockPolicy == ClockPolicy.Paused;

    public bool GameplayActive => ActiveGameplayPolicies.Contains(ClockPolicy);

    public bool CityActive => ActiveCityPolicies.Contains(ClockPolicy);

    public SimulationFrameStats LastFrame { get; private set; } =
        new() { PhysicsSteps = 0, CityTicks = 0, CityTimeScale = 0 };

    public ClockPolicy SetClockPolicy(ClockPolicy clockPolicy)
    {
        if (!Enum.IsDefined(clockPolicy))
        {
            throw new ArgumentOutOfRangeException(nameof(clockPolicy));
        }
        ClockPolicy = clockPolicy;
        return ClockPolicy;
    }

    public Func<bool> RegisterTask(
        string id,
        SimulationStage stage,
        Action<double, SimulationFrameContext> update,
        double order = 0,
        Func<SimulationFrameContext, bool>? enabled = null)
    {
        if (string.IsNullOrWhiteSpace(id)) throw new ArgumentException("Task ID is required.", nameof(id));
        if (taskIds.Contains(id)) throw new InvalidOperationException($"Simulation task already registered: {id}");
        if (!Enum.IsDefined(stage)) throw new ArgumentOutOfRangeException(nameof(stage));
        ArgumentNullException.ThrowIfNull(update);
        if (!double.IsFinite(order)) throw new ArgumentOutOfRangeException(nameof(order));

        var task = new ScheduledTask(id, update, order, enabled, registrationSerial++);
        List<ScheduledTask> stageTasks = tasks[stage];
        taskIds.Add(id);
        stageTasks.Add(task);
        stageTasks.Sort((left, right) =>
        {
            int comparison = left.Order.CompareTo(right.Order);
            return comparison != 0 ? comparison : left.Serial.CompareTo(right.Serial);
        });

        bool registered = true;
        return () =>
        {
            if (!registered) return false;
            registered = false;
            taskIds.Remove(id);
            return stageTasks.Remove(task);
        };
    }

    public double GetCityTimeScaleForPolicy()
    {
        if (!CityActive) return 0;
        double requested = getCityTimeScale(new ClockScaleContext(ClockPolicy, Frame));
        double safeScale = double.IsFinite(requested) ? Math.Max(0, requested) : 0;
        return MultipliedCityPolicies.Contains(ClockPolicy) ? safeScale : safeScale > 0 ? 1 : 0;
    }

    public SimulationFrameStats RunFrame(double timestampMs)
    {
        if (!double.IsFinite(timestampMs)) return Advance(0);
        if (!lastTimestampMs.HasValue)
        {
            lastTimestampMs = timestampMs;
            return Advance(0);
        }
        double deltaSeconds = Math.Max(0, (timestampMs - lastTimestampMs.Value) / 1000);
        lastTimestampMs = timestampMs;
        return Advance(deltaSeconds);
    }

    public void ResetFrameClock(double? timestampMs = null) =>
        lastTimestampMs = timestampMs.HasValue && double.IsFinite(timestampMs.Value) ? timestampMs : null;

    public SimulationSchedulerSnapshot AdvanceFrame(double realDeltaSeconds)
    {
        Advance(realDeltaSeconds);
        return Snapshot();
    }

    public SimulationSchedulerSnapshot Snapshot() => new()
    {
        Frame = Frame,
        ClockPolicy = ClockPolicy.ToToken(),
        Paused = Paused,
        GameplayActive = GameplayActive,
        CityActive = CityActive,
        FixedPhysicsStep = FixedPhysicsStep,
        CityStep = CityStep,
        PhysicsRemainder = physicsAccumulator,
        CityRemainder = cityAccumulator,
        LastFrame = LastFrame,
        Clocks = new ReadOnlyDictionary<string, SimulationClockSnapshot>(SnapshotClocks()),
    };

    private SimulationFrameStats Advance(double realDeltaSeconds)
    {
        double renderDelta = double.IsFinite(realDeltaSeconds) ? Math.Max(0, realDeltaSeconds) : 0;
        double boundedDelta = Math.Min(renderDelta, MaxFrameDelta);
        Frame++;
        AdvanceClock(SimulationScheduleCatalog.RenderClock, renderDelta);
        AdvanceClock(SimulationScheduleCatalog.UiClock, boundedDelta);

        var context = new SimulationFrameContext(() =>
            new ReadOnlyDictionary<string, SimulationClockSnapshot>(SnapshotClocks()))
        {
            Frame = Frame,
            ClockPolicy = ClockPolicy,
            RenderDelta = renderDelta,
            BoundedDelta = boundedDelta,
        };
        RunStage(SimulationStage.Input, boundedDelta, context);

        double gameplayDelta = GameplayActive ? boundedDelta : 0;
        double cityTimeScale = GetCityTimeScaleForPolicy();
        context.ClockPolicy = ClockPolicy;
        context.GameplayDelta = gameplayDelta;
        context.CityTimeScale = cityTimeScale;
        AdvanceClock(SimulationScheduleCatalog.GameplayClock, gameplayDelta);
        AdvanceClock(SimulationScheduleCatalog.PausedClock, Paused ? renderDelta : 0);
        clocks[SimulationScheduleCatalog.PhysicsClock].Delta = 0;
        clocks[SimulationScheduleCatalog.CityClock].Delta = 0;

        int physicsSteps = 0;
        if (gameplayDelta > 0)
        {
            physicsAccumulator += gameplayDelta;
            while (physicsAccumulator + Epsilon >= FixedPhysicsStep && physicsSteps < MaxPhysicsStepsPerFrame)
            {
                AdvanceClock(SimulationScheduleCatalog.PhysicsClock, FixedPhysicsStep);
                RunStage(SimulationStage.FixedPhysics, FixedPhysicsStep, context);
                physicsAccumulator -= FixedPhysicsStep;
                if (Math.Abs(physicsAccumulator) < Epsilon) physicsAccumulator = 0;
                physicsSteps++;
            }
        }
        context.PhysicsAlpha = Math.Min(1, physicsAccumulator / FixedPhysicsStep);

        if (gameplayDelta > 0) RunStage(SimulationStage.Gameplay, gameplayDelta, context);

        int cityTicks = 0;
        double cityDelta = gameplayDelta * cityTimeScale;
        if (cityDelta > 0)
        {
            cityAccumulator += cityDelta;
            while (cityAccumulator + Epsilon >= CityStep && cityTicks < MaxCityTicksPerFrame)
            {
                AdvanceClock(SimulationScheduleCatalog.CityClock, CityStep);
                RunStage(SimulationStage.City, CityStep, context);
                cityAccumulator -= CityStep;
                if (Math.Abs(cityAccumulator) < Epsilon) cityAccumulator = 0;
                cityTicks++;
            }
        }

        RunStage(SimulationStage.Presentation, boundedDelta, context);
        RunStage(SimulationStage.Camera, boundedDelta, context);
        RunStage(SimulationStage.Render, renderDelta, context);
        LastFrame = new SimulationFrameStats
        {
            PhysicsSteps = physicsSteps,
            CityTicks = cityTicks,
            CityTimeScale = cityTimeScale,
        };
        return LastFrame;
    }

    private void AdvanceClock(string name, double delta)
    {
        MutableClock clock = clocks[name];
        clock.Delta = delta;
        if (delta <= 0) return;
        clock.Elapsed += delta;
        clock.Ticks++;
    }

    private void RunStage(SimulationStage stage, double delta, SimulationFrameContext context)
    {
        context.Stage = stage;
        context.Clock = SimulationScheduleCatalog.GetClock(stage);
        context.Delta = delta;
        foreach (ScheduledTask task in tasks[stage].ToArray())
        {
            if (task.Enabled is not null && !task.Enabled(context)) continue;
            task.Update(delta, context);
        }
    }

    private Dictionary<string, SimulationClockSnapshot> SnapshotClocks() => clocks.ToDictionary(
        entry => entry.Key,
        entry => new SimulationClockSnapshot
        {
            Delta = entry.Value.Delta,
            Elapsed = entry.Value.Elapsed,
            Ticks = entry.Value.Ticks,
        },
        StringComparer.Ordinal);

    private static double RequirePositive(double value, string name) =>
        double.IsFinite(value) && value > 0 ? value : throw new ArgumentOutOfRangeException(name);

    private sealed class MutableClock
    {
        public double Delta { get; set; }
        public double Elapsed { get; set; }
        public long Ticks { get; set; }
    }

    private sealed record ScheduledTask(
        string Id,
        Action<double, SimulationFrameContext> Update,
        double Order,
        Func<SimulationFrameContext, bool>? Enabled,
        long Serial);
}

public sealed record ClockScaleContext(ClockPolicy ClockPolicy, long Frame);
